### PR TITLE
Add projections to layer context menu, allow grouping and nesting of menu items

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR adds 
A) convenience access to various intensity projection modes (max, min, mean, etc...)
B) an extended API for grouping and nesting items in the layer context menu actions dict (as requested/suggested by @goanpeca  in https://github.com/napari/napari/pull/2556#issuecomment-869983150) 

https://user-images.githubusercontent.com/1609449/125329011-fc7e8a80-e312-11eb-85e3-156fb44d55ec.mov

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by alisterburt** [comment]
I like the look of this! Had a play, the functionality is awesome and the API for grouping/nesting menu items looks great

**Comment by sofroniewn** [comment]
Looking cool!

File: napari/layers/_layer_actions.py [hunk 1]
```diff
@@ -42,8 +44,20 @@ def _split_stack(ll: LayerList, axis: int = 0):
...
+    layer = ll.selection.active
+    if not layer:
+        return
+
+    new = Layer.create(
+        getattr(np, mode)(layer.data, axis=axis),
+        {'name': f'{layer} {mode}-proj'},
+        layer._type_string,
+    )
+    ll.append(new)
```
Comment (jni) [inline napari/layers/_layer_actions.py]: @tlambert03 this logic won't work for layers based on coordinates. For example, for points layers, projecting along the ith dimension requires setting `layer.data[:, axis] = constant`. For surface layers, you would need to squish the `vertices` array in a similar way but leave `faces` untouched.

Comment (jni) [reply napari/layers/_layer_actions.py]: Also, for some layer types, only certain projection modes make sense. For example, you wouldn't want to do a mean projection on labels, so I would suggest not offering it. For points, only projection makes sense, I don't think any "value" projections make sense, *unless* you're going to be testing equality between floating point values and handling properties... Which you probably don't want to be doing in this PR!

**Review by jni — APPROVED** [review]
@tlambert03 I absolutely love where this is going, but I have a couple of concerns/issues:

1. As I've noted in the code, projecting over points is very different from projecting over an image. So that code needs to be removed (make this action only work for Image layers), or adapted.
2. I think projections are really a property of an *axis* more than a layer. Alternatively, if we want to keep the option to only project some layers but not all, we can do right click on layer, project > mode > axis/axes (in a dialog), or, right click on an *axis* (slider), project, mode, and either select a layer or apply to all layers.
3. Finally, you should use `np.ufunc(..., axis=i, keepdims=True)` to make sure that we keep a singleton dim for that axis, which will ensure correct broadcasting for higher-dimensional arrays, when projecting along an axis other than 0. (e.g. we might want to z-project in a 3D+t time series.)

Obviously none of these are show stoppers, and you can postpone almost all of them by limiting the PR to only work on image layers!

Overall though this is a super exciting addition. Lots of people will love this.

**Comment by tlambert03** [comment]
> I think projections are really a property of an axis more than a layer. Alternatively, if we want to keep the option to only project some layers but not all, we can do right click on layer, project > mode > axis/axes (in a dialog), or, right click on an axis (slider), project, mode, and either select a layer or apply to all layers.

projections are a property of a _both_ an axis and a layer, no?  (we certainly don't want to be forced into projecting _all_ layers in the layerlist, hence right clicking on a certain layer feels like a natural entry point..).  I feel like a pretty major use case is going to be the equivalent of the "z-stack" projections in fiji, so I do think we want to make that use case easy somehow, while still being able to do the completely arbitrary/complex projections.

what would you think of an additional item in the list ("custom") that opens a dialog that lists all of the axes available in the layer (along with their shape), and lets you check the ones you want to project?

File: napari/layers/_layer_actions.py (hunk 1 above)
Comment (tlambert03) [reply napari/layers/_layer_actions.py]: yep, will restrict for now.

**Comment by tlambert03** [comment]
OK, just looked through this again.  Would love to get this basic implementation in for the next version.
To address @jni's comments:

> projecting over points is very different from projecting over an image

dropped support for everything but images for now

> I think projections are really a property of an axis more than a layer. Alternatively, if we want to keep the option to only project some layers but not all, we can do right click on layer, project > mode > axis/axes (in a dialog)

i think it's a property of both an axis and a layer... and I suspect that most people will naturally gravitate to the layer list (rather than the axis labels) when they want to do a stack projection (or at least, I think it should also be possible).  Unfortunately, pretty much all of the axis information still lives on the viewer... which doesn't get passed at all to the layer list context menu (which is what this PR about), so it's currently quite hard to provide a dialog that even provides axis labels for someone to pick from (sure, we could do axis "0", "1", "2") ... but it's indicative of a scope problem, and I think that feature can safely come in another PR.  I suspect that _many_ people will benefit from this basic first-axis projection, this doesn't preclude us from expanding this with a more complicated interactive dialog in the future, and it also doesn't expose any public (non GUI) API that we might regret.

> np.ufunc(..., axis=i, keepdims=True)

did this... I did it a while back and I recall seeing something weird with it.  But I can't find it anymore... so we should just keep an eye out for it.  One slightly weird aspect is that objects with `ndim>2` but `np.squeeze().nim <= 2` will still get the "projection" menu enabled and allow projection.  To fix that, we'd need to include even more complexity in the "context" concept.  perhaps in the future...

**Review by sofroniewn — APPROVED** [review]
Ok, nice thanks for making these changes, I think we're now good to go! Will give this a merge. I'm pretty sure we're going to keep iterating on all this though, exited to see where it all goes!!

**Pull request merged**